processor/vote.go

```diff
@@ -0,0 +1,172 @@
+	stg    *storage.Storage
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// NewVoteProcessor creates a new VoteProcessor instance with the given storage
+// instance.
+func NewVoteProcessor(stg *storage.Storage) *VoteProcessor {
+	return &VoteProcessor{
+		stg: stg,
+	}
+}
+
+// Start method starts the vote processor. It will process ballots in the
+// background. It iterates over the ballots available in the storage and
+// generates proofs of the validity of the ballots, storing them back in the
+// storage. It will stop processing ballots when the context is cancelled.
+func (p *VoteProcessor) Start(ctx context.Context) error {
+	p.ctx, p.cancel = context.WithCancel(ctx)
+	// run the verifier in background until its context is cancelled
+	go func() {
+		for {
+			select {
+			case <-p.ctx.Done():
+				return
+			default:
+				// get the next ballot
+				ballot, key, err := p.stg.NextBallot()
+				if err != nil {
+					if err != storage.ErrNoMoreElements {
+						log.Errorf("failed to get next ballot: %v", err)
+					}
+					break
+				}
+				log.Debugf("new ballot to process for address %s", ballot.Address.String())
+				// process the ballot
+				verifiedBallot, err := p.ProcessBallot(ballot)
+				if err != nil {
+					log.Errorf("failed to process ballot: %v", err)
+					break
+				}
+				log.Debugf("ballot processed for address %s", ballot.Address.String())
+				// store the verified ballot
+				if err := p.stg.MarkBallotDone(key, verifiedBallot); err != nil {
+					log.Errorf("failed to mark ballot done: %v", err)
+				}
+			}
+			time.Sleep(1 * time.Second)
+		}
+	}()
+	return nil
+}
+
+// Stop method cancels the context of the vote processor, stopping the
+// processing of ballots.
+func (p *VoteProcessor) Stop() error {
+	p.cancel()
+	return nil
+}
+
+func (p *VoteProcessor) ProcessBallot(b *storage.Ballot) (*storage.VerifiedBallot, error) {
+	// decode the process id
+	pid := new(types.ProcessID)
+	if err := pid.Unmarshal(b.ProcessID); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal process id: %w", err)
+	}
+	// get the process metadata
+	process, err := p.stg.Process(pid)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get process metadata: %w", err)
+	}
+	// transform to circuit types
+	processID := crypto.BigToFF(circuits.BallotProofCurve.ScalarField(), b.ProcessID.BigInt().MathBigInt())
+	root := arbo.BytesToBigInt(process.Census.CensusRoot)
+	ballotMode := circuits.BallotModeToCircuit(*process.BallotMode)
+	encryptionKey := circuits.EncryptionKeyToCircuit(*process.EncryptionKey)
+	// calculate inputs hash
+	hashInputs := []*big.Int{}
+	hashInputs = append(hashInputs, processID)
+	hashInputs = append(hashInputs, root)
+	hashInputs = append(hashInputs, ballotMode.Serialize()...)
+	hashInputs = append(hashInputs, encryptionKey.Serialize()...)
+	hashInputs = append(hashInputs, b.Address.BigInt().MathBigInt())
+	hashInputs = append(hashInputs, b.Commitment.BigInt().MathBigInt())
```

Comment:
We do assume `b` is correct here and contains all required fields (not nil). 

We must make this assumption explicit in the method documentation and assure that the caller implements the correct checks.